src/pudl/transform/ferc1.py

```diff
@@ -97,6 +101,138 @@ class Ferc1RenameColumns(TransformParams):
+    to total up various hierarchical accounting categories (hydraulic turbines ->
+    hydraulic production plant -> all production plant -> all electric utility
+    plant) though the categorical columns required for that aggregation are added
+    later.
+    """
+    suffixes = "|".join(params.value_types)
+    pat = r"(^.*)_(" + suffixes + r"$)"
+    # filter out any columns that don't match the pattern
+    df_out = df.set_index(params.idx_cols).filter(regex=pat)
+    # check if there are unexpected columns being dropped
+    dropped_cols = [col for col in df if col not in df_out.reset_index()]
+    logger.debug(f"dropping: {dropped_cols}")
+    if params.expected_drop_cols != len(dropped_cols):
+        raise AssertionError(
+            f"Dropping more columns ({len(dropped_cols)}) than expected "
+            f"({params.expected_drop_cols}). Columns dropped: {dropped_cols}"
+        )
+
+    new_cols = pd.MultiIndex.from_tuples(
+        [(re.sub(pat, r"\1", col), re.sub(pat, r"\2", col)) for col in df_out.columns],
+        names=["xbrl_factoid", "value_type"],
+    )
+    df_out.columns = new_cols
+    df_out = (
+        df_out.stack(level="xbrl_factoid", dropna=False)
+        .loc[:, params.value_types]
+        .reset_index()
+    )
+    # remove the name of the columns which was the name of the renaming layer of the
+    # multi-index
+    df_out.columns.name = None
+    return df_out
+
+
+class MergeMetadata(TransformParams):
+    """Parameters for merging metadata."""
+
+    rename_columns: dict[str, str] = {}
+    """Dictionary to rename columns in the normalized metadata before merging.
+
+    This dictionary will be passed as :func:`pd.DataFrame.rename` ``columns`` parameter."""
+
+    on: str | None = None
+    """Column name to merge on in :func:`merge_metadata`."""
+
+
+def merge_metadata(
+    df: pd.DataFrame, xbrl_metadata_normalized: pd.DataFrame, params: MergeMetadata
+) -> pd.DataFrame:
+    """Merge metadata based on params."""
+    return pd.merge(
+        df,
+        xbrl_metadata_normalized.rename(columns=params.rename_columns),
+        on=params.on,
+        how="left",
+        validate="many_to_one",
+    )
+
+
+class AlignRowNumbersDbf(TransformParams):
+    """Parameters for aligning DBF row numbers with metadata from mannual maps."""
+
+    align_row_numbers_dbf: bool = False
+    """Boolean variable that indicates whether to run :func:`align_row_numbers_dbf`.
+
+    Default is ``False``.
+    """
+
+
+def align_row_numbers_dbf(df: pd.DataFrame, dbf_table_name: str) -> pd.DataFrame:
+    """Rename the xbrl_factoid column after :meth:`align_row_numbers_dbf`.
+
+    Note: This function takes a ``dbf_table_name`` because it needs to know which piece
+    of the dbf_to_xbrl map tp grab from :func:`read_dbf_to_xbrl_map`. This is difference
+    than our current general convention of passing in just a ``df`` and ``params``.
+    """
```

Comment:
What's the reasoning for keeping the `dbf_table_name` out of the parameters?

I'm not convinced this should be broken out as a function rather than only being a method, since it doesn't really have any parameters, except for an attribute which is part of the class it's embedded within, which it no longer has access to by virtue of being pulled out of the class as a function here.

We have lots of table specific methods that aren't parameterized elsewhere don't we? Like in the fuel table? What makes this one different from those? Is it just that it applies to a group of tables, rather than a single table?

Reply:
yeah i wasn't quite sure where to put this one for some these reasons. I strongly support having some boolean param or otherwise indicator of whether or not to apply this transform to a table in general. this enables us to put transforms into the standard  transform classes without having to make a bunch of sub-classes.

but that could be accomplished with the transform param and transform method (w/o a function that gets wrapped in the method).